File: src/main/java/skyblockclient/features/GhostBlock.java
package skyblockclient.features;

import net.minecraft.block.Block;
import net.minecraft.client.Minecraft;
import net.minecraft.init.Blocks;
import net.minecraft.item.ItemStack;
import net.minecraft.util.BlockPos;
import net.minecraftforge.event.entity.player.PlayerInteractEvent;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.gameevent.TickEvent;
import skyblockclient.SkyblockClient;
import skyblockclient.utils.SkyblockCheck;

import java.util.ArrayList;
import java.util.Arrays;

public class GhostBlock {

    private static final Minecraft mc = Minecraft.getMinecraft();
    private static final ArrayList<Block> interactables = new ArrayList<>(Arrays.asList(
            Blocks.acacia_door,
            Blocks.anvil,
            Blocks.beacon,
            Blocks.bed,
            Blocks.birch_door,
            Blocks.brewing_stand,
            Blocks.command_block,
            Blocks.crafting_table,
            Blocks.chest,
            Blocks.dark_oak_door,
            Blocks.daylight_detector,
            Blocks.daylight_detector_inverted,
            Blocks.dispenser,
            Blocks.dropper,
            Blocks.enchanting_table,
            Blocks.ender_chest,
            Blocks.furnace,
            Blocks.hopper,
            Blocks.jungle_door,
            Blocks.lever,
            Blocks.noteblock,
            Blocks.powered_comparator,
            Blocks.unpowered_comparator,
            Blocks.powered_repeater,
            Blocks.unpowered_repeater,
            Blocks.standing_sign,
            Blocks.wall_sign,
            Blocks.trapdoor,
            Blocks.trapped_chest,
            Blocks.wooden_button,
            Blocks.stone_button,
            Blocks.oak_door,
            Blocks.skull
    ));

    @SubscribeEvent
    public void tick(TickEvent.ClientTickEvent event) {
        if (!SkyblockCheck.inSkyblock || event.phase != TickEvent.Phase.START) return;
        if (SkyblockClient.keyBinds[2].isKeyDown()) {
            toAir(mc.objectMouseOver.getBlockPos());
        }
    }

    @SubscribeEvent
    public void playerInteract(PlayerInteractEvent event) {
        if (!SkyblockCheck.inSkyblock || !SkyblockClient.config.stonkGhostBlock) return;
        if (event.action == PlayerInteractEvent.Action.RIGHT_CLICK_BLOCK) {
            ItemStack item = mc.thePlayer.getHeldItem();
            if (item == null) return;
            if (item.getDisplayName().contains("Stonk")) {
                toAir(mc.objectMouseOver.getBlockPos());
                event.setCanceled(true);
            }
        }
    }

    private void toAir(BlockPos blockPos) {
        if (blockPos == null) return;
        Block block = mc.theWorld.getBlockState(mc.objectMouseOver.getBlockPos()).getBlock();
        if (!interactables.contains(block)) {
            mc.theWorld.setBlockToAir(mc.objectMouseOver.getBlockPos());
        }
    }

}

File: src/main/kotlin/skyblockclient/features/GhostBlock.kt
package skyblockclient.features

import net.minecraft.init.Blocks
import net.minecraft.item.ItemPickaxe
import net.minecraft.util.BlockPos
import net.minecraft.util.MovingObjectPosition
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import net.minecraftforge.fml.common.gameevent.TickEvent
import net.minecraftforge.fml.common.gameevent.TickEvent.ClientTickEvent
import skyblockclient.SkyblockClient.Companion.config
import skyblockclient.SkyblockClient.Companion.keyBinds
import skyblockclient.SkyblockClient.Companion.mc
import skyblockclient.events.ClickEvent
import skyblockclient.utils.LocationUtils.inSkyblock

object GhostBlock {

    private val blacklist = listOf(
        Blocks.acacia_door,
        Blocks.anvil,
        Blocks.beacon,
        Blocks.bed,
        Blocks.birch_door,
        Blocks.brewing_stand,
        Blocks.brown_mushroom,
        Blocks.chest,
        Blocks.command_block,
        Blocks.crafting_table,
        Blocks.dark_oak_door,
        Blocks.daylight_detector,
        Blocks.daylight_detector_inverted,
        Blocks.dispenser,
        Blocks.dropper,
        Blocks.enchanting_table,
        Blocks.ender_chest,
        Blocks.furnace,
        Blocks.hopper,
        Blocks.jungle_door,
        Blocks.lever,
        Blocks.noteblock,
        Blocks.oak_door,
        Blocks.powered_comparator,
        Blocks.powered_repeater,
        Blocks.red_mushroom,
        Blocks.skull,
        Blocks.standing_sign,
        Blocks.stone_button,
        Blocks.trapdoor,
        Blocks.trapped_chest,
        Blocks.unpowered_comparator,
        Blocks.unpowered_repeater,
        Blocks.wall_sign,
        Blocks.wooden_button
    )

    @SubscribeEvent
    fun onTick(event: ClientTickEvent) {
        if (event.phase != TickEvent.Phase.START || !inSkyblock || !keyBinds[2].isKeyDown) return
        toAir(mc.objectMouseOver.blockPos)
    }

    @SubscribeEvent
    fun onRightClick(event: ClickEvent.RightClickEvent) {
        if (!config.stonkGhostBlock || !inSkyblock || mc.objectMouseOver == null) return
        if (mc.objectMouseOver.typeOfHit == MovingObjectPosition.MovingObjectType.BLOCK) {
            if (mc.thePlayer.heldItem?.item is ItemPickaxe) {
                event.isCanceled = toAir(mc.objectMouseOver.blockPos)
            }
        }
    }

    private fun toAir(blockPos: BlockPos?): Boolean {
        if (blockPos != null) {
            val block = mc.theWorld.getBlockState(mc.objectMouseOver.blockPos).block
            if (!blacklist.contains(block)) {
                mc.theWorld.setBlockToAir(mc.objectMouseOver.blockPos)
                return true
            }
        }
        return false
    }
}
